flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/PulsarWriter.java

```diff
@@ -205,6 +209,12 @@ public void write(IN element, Context context) throws IOException, InterruptedEx
             builder.disableReplication();
         }
 
+        // Message Delay
+        long deliverAt = messageDelayer.deliverAt(element, context);
+        if (deliverAt > 0) {
+            builder.deliverAt(deliverAt);
```

Comment:
How does the message delayer work with flushing on checkpoint? What is the usual motivation behind that feature with Pulsar?

Reply:
Message delayer would only add an extra `sending time` property. It doesn't change any behavior of the message sending. A message with delayer would be normally sent, flushed on the checkpoint, and so on. We add this feature for some user-specified requirements, [message delay delivery](https://pulsar.apache.org/docs/en/next/concepts-messaging/#delayed-message-delivery) is wildly used in some companies like Baidu.